### DISCUSSION TITLE
Train a custom dataset with PyThorch Hub

### DISCUSSION THREAD

**Discussion by alevangel** [ultralytics/yolov5#4631] [Q&A]
I know that for training a custom dataset with train.py there is an extensive documentation, but how to dao that with PyTorch? 

I only see from the [doc on PyTorch ](https://github.com/ultralytics/yolov5/issues/36) that I have have to use this line:
`model = torch.hub.load('ultralytics/yolov5', 'yolov5m', autoshape=False)`

But then how to train the model?

**Comment by glenn-jocher** [comment]
@alevangel the command above will produce a pretrained YOLOv5m model, you provide the training code to train it.

See **Train Custom Data** tutorial below for a fully managed training solution.

## YOLOv5 Tutorials

* [Train Custom Data](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://github.com/ultralytics/yolov5/wiki/Tips-for-Best-Training-Results)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://github.com/ultralytics/yolov5/issues/2518)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://github.com/ultralytics/yolov5/issues/475)
* [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36)&nbsp; ⭐ NEW
* [TorchScript, ONNX, CoreML Export](https://github.com/ultralytics/yolov5/issues/251) 🚀
* [Test-Time Augmentation (TTA)](https://github.com/ultralytics/yolov5/issues/303)
* [Model Ensembling](https://github.com/ultralytics/yolov5/issues/318)
* [Model Pruning/Sparsity](https://github.com/ultralytics/yolov5/issues/304)
* [Hyperparameter Evolution](https://github.com/ultralytics/yolov5/issues/607)
* [Transfer Learning with Frozen Layers](https://github.com/ultralytics/yolov5/issues/1314)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Reply by alevangel** [reply]
But I only see train a custom dataset with a notebook using the train.py, 
is possible to do that directly in PyCharm for istance?

Sorry if this is a dumb question

**Reply by glenn-jocher** [reply]
@alevangel PyCharm is a python IDE, naturally it's capable of everything a notebook can do.

**Reply by GTimothee** [reply]
@alevangel in this repo yolov5 gives you a train.py script that embed all the training functions etc. 

When you call train.py the script loads the model, loads the dataset and execute a training function. 

If you want to do it yourself, you can load the model yourself using pytorch hub as you mentioned, and then you have to load your dataset yourself, write the training function etc. 

For more information and if you are a complete beginner, I would suggest [following this tutorial](https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html). It explains how to fine tune a pytorch model. Then , just use the model defined with ```model = torch.hub.load('ultralytics/yolov5', 'yolov5m', autoshape=False)``` in the training loop of the tutorial. 

Hope it helps.

**Reply by alevangel** [reply]
@GTimothee thanks a lot for this answer! Is what I was looking for, it helped a lot!
